File: go/auth_token_manager.go
package modal

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"sync/atomic"
	"time"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"golang.org/x/sync/singleflight"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

const (
	// Start refreshing this many seconds before the token expires
	RefreshWindow       = 5 * 60
	authTokenGetTimeout = 5 * time.Second
	// Bounds a whole shared refresh, whose retries are detached from any caller's context and so otherwise unlimited.
	// Deliberately shorter than the full retry budget of authTokenGetTimeout per attempt: a refresh that has been
	// failing this long is better given up on, and callers with a cached token keep using it. Matches the Python
	// client's AUTH_TOKEN_RETRY_TOTAL_TIMEOUT.
	authTokenRefreshTimeout = 3 * authTokenGetTimeout
	// After a failed refresh, wait before hitting the server again, growing exponentially
	// with consecutive failures between these bounds.
	failureBackoffBase = 500 * time.Millisecond
	failureBackoffMax  = 60 * time.Second
	// If the token doesn't have an expiry field, default to current time plus this value (not expected).
	DefaultExpiryOffset = 20 * 60
)

type tokenAndExpiry struct {
	token  string
	expiry int64
}

type atomicDuration struct {
	nanos atomic.Int64
}

func (d *atomicDuration) Load() time.Duration {
	return time.Duration(d.nanos.Load())
}

func (d *atomicDuration) Store(v time.Duration) {
	d.nanos.Store(int64(v))
}

// authTokenManager manages authentication tokens, refreshing them lazily
// when GetToken is called. Tokens are refreshed when expired or within
// RefreshWindow seconds of expiry.
type authTokenManager struct {
	client pb.ModalClientClient
	logger *slog.Logger

	tokenAndExpiry atomic.Pointer[tokenAndExpiry]
	refreshGroup   singleflight.Group
	// Claimed by the caller that takes on a pre-expiry refresh, so the others keep using the still-valid token
	// instead of waiting for the request.
	refreshing atomic.Bool
	retryAfter atomic.Pointer[time.Time]
	backoff    atomicDuration
}

func newAuthTokenManager(client pb.ModalClientClient, logger *slog.Logger) *authTokenManager {
	manager := &authTokenManager{
		client: client,
		logger: logger,
	}

	manager.tokenAndExpiry.Store(&tokenAndExpiry{
		token:  "",
		expiry: 0,
	})
	manager.backoff.Store(failureBackoffBase)

	return manager
}

// GetToken returns a valid auth token, fetching or refreshing as needed.
//
// Three states:
//  1. Valid token (not near expiry): returned immediately, no locking.
//  2. No token or expired: all callers block until a fresh token is fetched.
//     Only one goroutine makes the RPC; the others wait for its outcome.
//  3. Valid but within RefreshWindow of expiry: one goroutine refreshes
//     (blocking only itself); concurrent callers get the old, still-valid token.
func (m *authTokenManager) GetToken(ctx context.Context) (string, error) {
	data := m.tokenAndExpiry.Load()

	if data.token == "" || isExpired(*data) {
		return m.tryRefreshToken(ctx)
	}

	if m.shouldRefresh(*data) && m.refreshing.CompareAndSwap(false, true) {
		results := m.startSharedRefresh(ctx)
		var token string
		var err error
		select {
		case result := <-results:
			m.refreshing.Store(false)
			token, err = tokenFromRefresh(result)
		case <-ctx.Done():
			// The refresh outlives this caller, so keep it claimed until it finishes: later callers should keep using
			// the still-valid token instead of waiting on a request they don't need.
			go func() {
				<-results
				m.refreshing.Store(false)
			}()
			err = ctx.Err()
		}
		if err != nil {
			if isAuthDenied(err) {
				return "", err
			}
			m.logger.WarnContext(ctx, "Auth token refresh failed; falling back to cached token", "error", err)
			return data.token, nil
		}
		return token, nil
	}

	return data.token, nil
}

func (m *authTokenManager) tryRefreshToken(ctx context.Context) (string, error) {
	token, err := m.sharedRefreshToken(ctx)
	if err == nil {
		return token, nil
	}
	if isAuthDenied(err) {
		return "", err
	}
	data := m.tokenAndExpiry.Load()
	if data.token == "" {
		return "", err
	}
	m.logger.WarnContext(ctx, "Auth token refresh failed; falling back to cached token", "error", err)
	return data.token, nil
}

// sharedRefreshToken returns the token from a single in-flight refresh: the first caller makes the RPC while the
// others wait for its outcome, so a burst of callers results in one AuthTokenGet whether it succeeds or fails. The
// request is detached from the calling context, so a caller going away doesn't abort the refresh the remaining
// callers are waiting on, and bounded by authTokenRefreshTimeout so a stuck refresh can't block later ones.
func (m *authTokenManager) sharedRefreshToken(ctx context.Context) (string, error) {
	results := m.startSharedRefresh(ctx)
	select {
	case result := <-results:
		return tokenFromRefresh(result)
	case <-ctx.Done():
		return "", ctx.Err()
	}
}

// startSharedRefresh joins the in-flight refresh, starting one if there isn't one, and returns the channel its
// outcome is delivered on.
func (m *authTokenManager) startSharedRefresh(ctx context.Context) <-chan singleflight.Result {
	return m.refreshGroup.DoChan("refresh", func() (any, error) {
		// Maybe another goroutine refreshed while this call was being shared.
		if data := m.tokenAndExpiry.Load(); data.token != "" && !m.shouldRefresh(*data) {
			return data.token, nil
		}
		rpcCtx, cancel := context.WithTimeout(context.WithoutCancel(ctx), authTokenRefreshTimeout)
		defer cancel()
		return m.FetchToken(rpcCtx)
	})
}

func tokenFromRefresh(result singleflight.Result) (string, error) {
	if result.Err != nil {
		return "", result.Err
	}
	// singleflight predates generics, so the token comes back as an any.
	return result.Val.(string), nil
}

// FetchToken fetches a new token using AuthTokenGet() and stores it.
func (m *authTokenManager) FetchToken(ctx context.Context) (string, error) {
	retries := 0
	if m.tokenAndExpiry.Load().token == "" {
		// No cached token to fall back on, so a failure is user-visible: retry transient errors.
		// Otherwise one attempt, and retryAfter handles the cooldown.
		retries = defaultRetryAttempts
	}
	resp, err := m.client.AuthTokenGet(ctx, &pb.AuthTokenGetRequest{},
		timeoutCallOption{timeout: authTokenGetTimeout},
		retryCallOption{retries: &retries},
	)
	if err == nil && resp.GetToken() == "" {
		err = fmt.Errorf("internal error: did not receive auth token from server, please contact Modal support")
	}
	if err != nil {
		if isAuthDenied(err) {
			return "", err
		}
		// Back off (exponentially on consecutive failures) so we don't hammer a struggling server.
		cur := m.backoff.Load()
		t := time.Now().Add(cur)
		m.retryAfter.Store(&t)
		next := cur * 2
		if next > failureBackoffMax {
			next = failureBackoffMax
		}
		m.backoff.Store(next)
		return "", err
	}

	token := resp.GetToken()
	var expiry int64
	if exp := m.decodeJWT(token); exp > 0 {
		expiry = exp
	} else {
		m.logger.WarnContext(ctx, "x-modal-auth-token does not contain exp field")
		// We'll use the token, and set the expiry to 20 min from now.
		expiry = time.Now().Unix() + DefaultExpiryOffset
	}

	m.tokenAndExpiry.Store(&tokenAndExpiry{
		token:  token,
		expiry: expiry,
	})

	timeUntilRefresh := time.Duration(expiry-time.Now().Unix()-RefreshWindow) * time.Second
	m.logger.DebugContext(ctx, "Fetched auth token",
		"expires_in", time.Until(time.Unix(expiry, 0)),
		"refresh_in", timeUntilRefresh)

	m.retryAfter.Store(nil)
	m.backoff.Store(failureBackoffBase)
	return token, nil
}

// Extracts the exp claim from a JWT token.
func (m *authTokenManager) decodeJWT(token string) int64 {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return 0
	}

	payload := parts[1]
	for len(payload)%4 != 0 {
		payload += "="
	}

	decoded, err := base64.URLEncoding.DecodeString(payload)
	if err != nil {
		return 0
	}

	var claims map[string]interface{}
	if err := json.Unmarshal(decoded, &claims); err != nil {
		return 0
	}

	if exp, ok := claims["exp"].(float64); ok {
		return int64(exp)
	}

	return 0
}

// GetCurrentToken returns the current cached token.
func (m *authTokenManager) GetCurrentToken() string {
	return m.tokenAndExpiry.Load().token
}

// IsExpired checks if the current token is expired.
func (m *authTokenManager) IsExpired() bool {
	return isExpired(*m.tokenAndExpiry.Load())
}

func isExpired(data tokenAndExpiry) bool {
	return time.Now().Unix() >= data.expiry
}

func needsRefresh(data tokenAndExpiry) bool {
	return time.Now().Unix() >= data.expiry-RefreshWindow
}

func (m *authTokenManager) inBackoff() bool {
	retryAfter := m.retryAfter.Load()
	return retryAfter != nil && time.Now().Before(*retryAfter)
}

func (m *authTokenManager) shouldRefresh(data tokenAndExpiry) bool {
	return needsRefresh(data) && !m.inBackoff()
}

func isAuthDenied(err error) bool {
	// The credentials themselves were rejected (e.g. revoked/invalid key), as opposed to a
	// transient outage/overload. Fail fast instead of reusing the cached token or backing off.
	st, ok := status.FromError(err)
	return ok && (st.Code() == codes.Unauthenticated || st.Code() == codes.PermissionDenied)
}

// SetToken sets the token and expiry (for testing).
func (m *authTokenManager) SetToken(token string, expiry int64) {
	m.tokenAndExpiry.Store(&tokenAndExpiry{
		token:  token,
		expiry: expiry,
	})
}

File: js/src/auth_token_manager.ts
import { ClientError, Status } from "nice-grpc";
import type { Logger } from "./logger";

// Start refreshing this many seconds before the token expires
export const REFRESH_WINDOW = 5 * 60;
export const AUTH_TOKEN_GET_TIMEOUT_MS = 5 * 1000;
export const AUTH_TOKEN_GET_MAX_RETRIES = 3;
// After a failed refresh, wait before hitting the server again, growing exponentially with
// consecutive failures between these bounds (in milliseconds).
export const FAILURE_BACKOFF_BASE_MS = 500;
export const FAILURE_BACKOFF_MAX_MS = 60 * 1000;
// If the token doesn't have an expiry field, default to current time plus this value (not expected).
export const DEFAULT_EXPIRY_OFFSET = 20 * 60;

/**
 * Manages authentication tokens, refreshing them lazily when getToken is
 * called. Tokens are refreshed when expired or within REFRESH_WINDOW seconds
 * of expiry.
 *
 * Three states:
 *  1. Valid token (not near expiry): returned immediately.
 *  2. No token or expired: all callers block until a fresh token is fetched.
 *     Only one fetch happens; concurrent callers await the same promise.
 *  3. Valid but within REFRESH_WINDOW of expiry: if no refresh is in progress
 *     the caller triggers one (blocking itself); concurrent callers get the
 *     old, still-valid token.
 */
export class AuthTokenManager {
  private client: any;
  private logger: Logger;
  private currentToken: string = "";
  private tokenExpiry: number = 0;
  private refreshPromise: Promise<void> | null = null;
  private retryAfter: number = 0;
  private backoffMs: number = FAILURE_BACKOFF_BASE_MS;

  constructor(client: any, logger: Logger) {
    this.client = client;
    this.logger = logger;
  }

  /**
   * Returns a valid auth token.
   */
  async getToken(): Promise<string> {
    if (!this.currentToken || this.isExpired()) {
      return this.tryRefreshToken();
    }

    if (this.shouldRefresh() && !this.refreshPromise) {
      await this.tryRefreshToken();
    }

    return this.currentToken;
  }

  private async tryRefreshToken(): Promise<string> {
    try {
      return await this.lockedRefreshToken();
    } catch (error) {
      if (this.isAuthDenied(error)) {
        throw error;
      }
      if (!this.currentToken) {
        throw error;
      }
      this.logger.warn(
        "Auth token refresh failed; falling back to cached token",
        "error",
        error,
      );
      return this.currentToken;
    }
  }

  /**
   * Ensures only one fetch is in progress at a time. Concurrent callers
   * await the same promise. Includes a double-check so that if another
   * caller already refreshed, we skip the RPC.
   */
  private async lockedRefreshToken(): Promise<string> {
    if (this.currentToken && !this.shouldRefresh()) {
      return this.currentToken;
    }
    if (!this.refreshPromise) {
      this.refreshPromise = (async () => {
        try {
          await this.fetchToken();
        } finally {
          this.refreshPromise = null;
        }
      })();
    }
    await this.refreshPromise;
    return this.currentToken;
  }

  /**
   * Fetches a new auth token from the server and stores it.
   */
  private async fetchToken(): Promise<void> {
    try {
      const hasCachedToken = !!this.currentToken;
      const response = await this.client.authTokenGet(
        {},
        {
          // No cached token to fall back on, so a failure is user-visible: retry transient errors.
          // Otherwise one attempt, and retryAfter handles the cooldown.
          retries: hasCachedToken ? 0 : AUTH_TOKEN_GET_MAX_RETRIES,
          timeoutMs: AUTH_TOKEN_GET_TIMEOUT_MS,
        },
      );
      const token = response.token;
      if (!token) {
        throw new Error(
          "Internal error: did not receive auth token from server, please contact Modal support",
        );
      }
      this.currentToken = token;
      const exp = this.decodeJWT(token);
      if (exp > 0) {
        this.tokenExpiry = exp;
      } else {
        this.logger.warn("x-modal-auth-token does not contain exp field");
        // We'll use the token, and set the expiry to DEFAULT_EXPIRY_OFFSET from now.
        this.tokenExpiry =
          Math.floor(Date.now() / 1000) + DEFAULT_EXPIRY_OFFSET;
      }
      this.retryAfter = 0;
      this.backoffMs = FAILURE_BACKOFF_BASE_MS;

      const now = Math.floor(Date.now() / 1000);
      const expiresIn = this.tokenExpiry - now;
      const refreshIn = this.tokenExpiry - now - REFRESH_WINDOW;
      this.logger.debug(
        "Fetched auth token",
        "expires_in",
        `${expiresIn}s`,
        "refresh_in",
        `${refreshIn}s`,
      );
    } catch (error) {
      if (this.isAuthDenied(error)) {
        throw error;
      }
      // Back off (exponentially on consecutive failures) so we don't hammer a struggling server.
      this.retryAfter = Date.now() + this.backoffMs;
      this.backoffMs = Math.min(this.backoffMs * 2, FAILURE_BACKOFF_MAX_MS);
      throw error;
    }
  }

  /**
   * Extracts the exp claim from a JWT token.
   */
  private decodeJWT(token: string): number {
    try {
      const parts = token.split(".");
      if (parts.length !== 3) {
        return 0;
      }

      let payload = parts[1];
      while (payload.length % 4 !== 0) {
        payload += "=";
      }

      const decoded = atob(payload);
      const claims = JSON.parse(decoded);
      return claims.exp || 0;
    } catch {
      return 0;
    }
  }

  isExpired(): boolean {
    const now = Math.floor(Date.now() / 1000);
    return now >= this.tokenExpiry;
  }

  private needsRefresh(): boolean {
    const now = Math.floor(Date.now() / 1000);
    return now >= this.tokenExpiry - REFRESH_WINDOW;
  }

  private inBackoff(): boolean {
    return Date.now() < this.retryAfter;
  }

  private isAuthDenied(error: unknown): boolean {
    // The credentials themselves were rejected (e.g. revoked/invalid key), as opposed to a
    // transient outage/overload. Fail fast instead of reusing the cached token or backing off.
    return (
      error instanceof ClientError &&
      (error.code === Status.UNAUTHENTICATED ||
        error.code === Status.PERMISSION_DENIED)
    );
  }

  private shouldRefresh(): boolean {
    return this.needsRefresh() && !this.inBackoff();
  }

  getCurrentToken(): string {
    return this.currentToken;
  }

  setToken(token: string, expiry: number): void {
    this.currentToken = token;
    this.tokenExpiry = expiry;
  }
}
